### PR TITLE
Implement "How to write your Story" flow

### LINKED ISSUE #22891: [Project - Onboarding]: Implement flow: "How to write your first Story"
## Description

The goal of this task is to guide new users in writing their first Component Story Format (CSF) file in Storybook. After the user completes the initial step-by-step guide, a modal will appear, explaining the structure of a CSF file including the meta field, story exports, etc. In the final step, the user will be asked to copy a story and paste it into the button Story. Upon successful recognition of the new story, a "continue" button will lead the user to the new story, accompanied by a confetti rain animation to celebrate their achievement.

## Acceptance Criteria:

- After the first step-by-step guide is completed, a modal should appear, explaining how to create a CSF file, its structure, and its main components. It will explain the `import`, `meta`, `stories` and `args` part of a typical CSF file.
- The modal should guide the user to copy a story and paste it into the button Story.
- The system should correctly recognize the new story.
- A "Go to Story" button should appear upon successful recognition of the new story, which will lead the user to the new story.
- Upon navigating to the new story via the "Go to Story" button, a confetti rain animation should be triggered.
- All existing features and functionalities should remain unaffected.

## (To be discussed) Edge Cases to Consider:

- Incorrect CSF file structure: If the user does not correctly follow the CSF file structure, the system should provide a meaningful error message and guide the user to correct it.
- Failure to recognize the new story: The system should handle scenarios where it fails to recognize the new story, providing helpful error messages to guide the user.
- Navigation and animation triggers: Ensure the "Go to Story" button and the confetti rain animation trigger correctly and at the appropriate times.

### PR DESCRIPTION
**Pull request by valentinpalkovic**
Closes https://github.com/storybookjs/storybook/issues/22891

## What I did

I have implemented the "How to write your Story" flow.

<details>
  <summary>📦 Published PR as canary version: <code>0.0.19--canary.33.684ad81.0</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @storybook/addon-onboarding@0.0.19--canary.33.684ad81.0
  # or 
  yarn add @storybook/addon-onboarding@0.0.19--canary.33.684ad81.0
  ```
</details>

### CODE REVIEW

**Comment by yannbf** [comment]
@cdedreuille there is a slight text difference in the sidebar image example vs the real one, a missing "s":

image

image

Could you make another image with "example" instead? Same goes for the "title" part, it should be "Example/Button"

**Comment by yannbf** [comment]
@cdedreuille 
I believe this image should either contain `label: 'Click here'` or the control should just have "Click" as text:

image

**Comment by yannbf** [comment]
@cdedreuille the color of the boolean control in Storybook 7 is blueish rather than grey:
image

**Comment by yannbf** [comment]
@cdedreuille @valentinpalkovic what do you think of highlighting important concept names, either in `Code` notation or **Bold** in the sentences?

e.g.
image

would be something like:

The default export, **Meta**, contains metadata about this component's stories. 
The **title** field controls where stories appear in the sidebar.

**Comment by valentinpalkovic** [comment]
> @cdedreuille @valentinpalkovic what do you think of highlighting important concept names, either in `Code` notation or **Bold** in the sentences?
> 
> e.g. image
> 
> would be something like:
> 
> The default export, **Meta**, contains metadata about this component's stories. The **title** field controls where stories appear in the sidebar.

Makes sense. Up to @cdedreuille, though :)

File: src/features/WriteStoriesModal/WriteStoriesModal.tsx [hunk 1]
```diff
@@ -2,27 +2,272 @@ import React from "react";
...
-            steps below.
-          </Description>
-          <Button style={{ marginTop: 4 }} onClick={onFinish}>
-            Go to story
-          </Button>
-        </ModalContentWrapper>
+    <Modal width={738} defaultOpen>
+      {({ Title, Description: DefaultDescription, Close }) => (
+        <ModalContent>
+          <div style={{ maxHeight: "450px" }}>
```
Comment (yannbf) [inline src/features/WriteStoriesModal/WriteStoriesModal.tsx]: Should this contain a minHeight too, so that the layout jump is not too much?

Comment (valentinpalkovic) [reply src/features/WriteStoriesModal/WriteStoriesModal.tsx]: oh yeah! I forgot about it. I will take a look!

File: src/features/WriteStoriesModal/WriteStoriesModal.tsx [hunk 2]
```diff
@@ -2,27 +2,272 @@ import React from "react";
...
+                    ? 3
+                    : step === "args"
+                    ? 4
+                    : step === "customStory"
+                    ? 5
+                    : 1
+                }
+                contents={data}
+                width="445px"
+              />
```
Comment (yannbf) [inline src/features/WriteStoriesModal/WriteStoriesModal.tsx]: Up to preference, but you could do something like this instead
```suggestion
              const stepIndex = {
                  imports: 1,
                  meta: 2,
                  story: 3,
                  args: 4,
                  customStory: 5,
                }

              // ...
              <SyntaxHighlighter
                activeStep={stepIndex[step] || 1}
                contents={data}
                width="445px"
              />
```

File: src/features/WriteStoriesModal/WriteStoriesModal.tsx [hunk 3]
```diff
@@ -2,27 +2,272 @@ import React from "react";
...
+  addonsStore: AddonStore;
+}) {
+  const [step, setStep] = React.useState<
+    "imports" | "meta" | "story" | "args" | "customStory"
+  >("imports");
+
+  const [isWarningStoryCopied, setWarningStoryCopied] = React.useState(false);
+
+  const buttonPath = useGetButtonPath();
+  const doesWarningButtonExist = useGetWarningButtonStatus(
```
Comment (yannbf) [inline src/features/WriteStoriesModal/WriteStoriesModal.tsx]: maybe this should be called `warningButtonStatus` instead? `doesWarningButtonExist` sounds like it refers to a boolean value

File: src/features/WriteStoriesModal/WriteStoriesModal.tsx [hunk 4]
```diff
@@ -2,27 +2,272 @@ import React from "react";
...
+                  position: "absolute",
+                  top: backdropBoundary.top + backdropBoundary.height - 45,
+                  left:
+                    backdropBoundary.left +
+                    backdropBoundary.width -
+                    (isWarningStoryCopied ? 115 : 100),
+                  zIndex: 1000,
+                }}
+              >
+                {isWarningStoryCopied ? <>Code copied </> : "Copy code"}
```
Comment (yannbf) [inline src/features/WriteStoriesModal/WriteStoriesModal.tsx]: I don't know if there is enough space, but "Copied to clipboard" would be more accessible to users:

```suggestion
                {isWarningStoryCopied ? "Copied to clipboard" : "Copy code"}
```

File: src/features/WriteStoriesModal/WriteStoriesModal.tsx (hunk 2 above)
Comment (valentinpalkovic) [reply src/features/WriteStoriesModal/WriteStoriesModal.tsx]: I like this proposed change!

File: src/features/WriteStoriesModal/WriteStoriesModal.tsx (hunk 3 above)
Comment (valentinpalkovic) [reply src/features/WriteStoriesModal/WriteStoriesModal.tsx]: I agree

File: src/features/WriteStoriesModal/hooks/useGetWarningButtonStatus.tsx [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+          if (out) {
+            setStatus({ data: true, error: null });
+          }
+        });
+      };
+
+      addonsStore
+        .getServerChannel()
+        .on(STORY_INDEX_INVALIDATED, getWarningButtonStatus);
```
Comment (yannbf) [inline src/features/WriteStoriesModal/hooks/useGetWarningButtonStatus.tsx:30]: Today I learned 🤔

**Review by yannbf — APPROVED** [review]
LGTM!

**Pull request merged**